ext/pcntl/pcntl.c

```diff
@@ -1476,6 +1481,115 @@ PHP_FUNCTION(pcntl_setns)
 }
 #endif
 
+#ifdef HAVE_SCHED_SETAFFINITY
+PHP_FUNCTION(pcntl_getcpuaffinity)
+{
+	zend_long pid;
+	bool pid_is_null = 1;
+	cpu_set_t mask;
+
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_LONG_OR_NULL(pid, pid_is_null)
+	ZEND_PARSE_PARAMETERS_END();
+
+	// 0 == getpid in this context, we re just saving a syscall
+	pid = pid_is_null ? 0 : pid;
+
+	CPU_ZERO(&mask);
+
+	if (sched_getaffinity(pid, sizeof(mask), &mask) != 0) {
+		PCNTL_G(last_error) = errno;
+		switch (errno) {
+			case ESRCH:
+				zend_argument_value_error(1, "invalid process (" ZEND_LONG_FMT ")", pid);
+				RETURN_THROWS();
+			case EPERM:
+				php_error_docref(NULL, E_WARNING, "Calling process not having the proper privileges");
+				break;
+			default:
+				php_error_docref(NULL, E_WARNING, "Error %d", errno);
+		}
+
+		RETURN_EMPTY_ARRAY();
+	}
+
+	zend_ulong maxcpus = (zend_ulong)sysconf(_SC_NPROCESSORS_ONLN);
```

Comment:
Genuine question: can the number of CPUs be lower than the highest in-use CPU id?

Intuitively, I think that _could_ be possible. Let's say we have 8 CPU cores (0-7), and the admin disables core 0. Then the number of online CPU cores is 7 instead of 8, but the highest id is still 8?

Reply:
do you mean while the process is running ?

Reply:
Not necessarily. The situation I propose could also have occurred prior to running the script.
I don't know if that's actually how it works, but I would expect core numbers / CPU ids to remain unchanged.

Reply:
If that is prior sysconf(_SC_NPROCESSORS_ONLN) reflects the change then. Informations are stored into `/sys/devices/system/cpu/cpu<id>` files but thankfully we do not need to go through them :)

Reply:
I realize I'm not explaining my point well. I just tested what would happen and this is actually a problem.

I have 8 CPUs in my computer. Therefore, `sysconf(_SC_NPROCESSORS_ONLN)` is 8.
They are numbered 0-7.

Consider this program:
```c
#define _GNU_SOURCE
#include <err.h>
#include <sched.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main(int argc, char *argv[])
{
    cpu_set_t     set;
    CPU_ZERO(&set);

    CPU_SET(2, &set);

    printf("nr online %d\n", sysconf(_SC_NPROCESSORS_ONLN));

    if (sched_setaffinity(getpid(), sizeof(set), &set) == -1)
        err(EXIT_FAILURE, "sched_setaffinity");

    return 0;
}
```

This program will succeed, because CPU 2 is available. We also see that it prints `nr online 8`.

Now, I disable CPU number 2: `echo "0" > /sys/devices/system/cpu/cpu2/online`

Now I rerun the program, I see this:
```
nr online 7
a.out: sched_setaffinity: Invalid argument
```
This makes sense because we have one less core available, _and_ CPU nr 2 is now no longer a valid number, therefore the system call complains.
But, this also means that the valid CPU numbers are now: [0, 1, 3, 4, 5, 6, 7].

This is a problem both for this function, and the other function.

Looking at the code your wrote: note that now `maxcpus == 7`, and if the `mask` variable includes CPU number 7, it will _not_ be included in the `return_value` array because `i` will never become 7 in the for loop.
Take a look at this code for example:
```c
#define _GNU_SOURCE
#include <err.h>
#include <sched.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main(int argc, char *argv[])
{
    cpu_set_t     set;
    CPU_ZERO(&set);

    sched_getaffinity(0, sizeof(set), &set);

    int online = sysconf(_SC_NPROCESSORS_ONLN);
    printf("nr online %d\n", online);

    for (int i = 0; i < online; i++) {
        if (CPU_ISSET(i, &set)) {
            printf("on CPU %d\n", i);
        }
    }
    return 0;
}
```

This will output the following, after disabling CPU 2:
```
nr online 7
on CPU 0
on CPU 1
on CPU 3
on CPU 4
on CPU 5
on CPU 6
```
Notice CPU 7 is missing.